src/go/intelligence/pipeline/add_annotations.go

```diff
@@ -0,0 +1,139 @@
+package pipeline
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"go/intelligence/models"
+	"go/intelligence/utils"
+	"io/ioutil"
+	"log"
+	"net/http"
+	"net/http/httputil"
+	"strings"
+	"time"
+)
+
+// Module     : intelligence
+// Author     : Rahul Indra <indrarahul2013 AT gmail dot com>
+// Created    : Wed, 1 July 2020 11:04:01 GMT
+// Description: CMS MONIT infrastructure Intelligence Module
+
+//AddAnnotation function for adding annotations to dashboards
+func AddAnnotation(data <-chan models.AmJSON) <-chan models.AmJSON {
+
+	dataAfterAnnotation := make(chan models.AmJSON)
+
+	go func() {
+		dashboards := utils.FindDashboard()
```

Comment:
Here you don't need to finddashboard every single time since this information is static. Instead, you should add a cache and update the cache periodically, e.g. once a day. This will save time, traffic and number of calls to upstream server. For cache, please define a struct which will hold the dashboard map and expiration timestamp. Then add FindDatashboard method to the struct which will first check the cache, if it is expired it will use utils.FindDashboard to fetch new map, update the map, update the expiration time and return it.

Reply:
@vkuznet This will decrease our time compexity for sure. I have added caching feature for the dashboards.